posthog/temporal/data_imports/pipelines/meta_ads/source.py

```diff
@@ -0,0 +1,445 @@
+    """Get the Meta Ads integration."""
+    return Integration.objects.get(id=config.meta_ads_integration_id, team_id=team_id)
+
+
+# Simple schema structure without typing
+class MetaAdsSchema:
+    def __init__(self, name: str, primary_key: list[str], field_names: list[str]):
+        self.name = name
+        self.primary_key = primary_key
+        self.field_names = field_names
+
+
+# Note: can make this static but keeping schemas.py to match other schema files for now
+def get_schemas() -> dict[str, MetaAdsSchema]:
+    """Obtain Meta Ads schemas using predefined field definitions."""
+    schemas = {}
+
+    for resource_name, schema_def in RESOURCE_SCHEMAS.items():
+        field_names = schema_def["field_names"].copy()
+        primary_key = schema_def["primary_key"]
+
+        schema = MetaAdsSchema(
+            name=resource_name,
+            primary_key=primary_key,
+            field_names=field_names,
+        )
+        schemas[resource_name] = schema
+
+    return schemas
+
+
+def _serialize_value(value: Any) -> Any:
+    """Serialize complex values to JSON strings."""
+    if isinstance(value, dict | list):
+        return json.dumps(value)
+    return value
+
+
+def _make_api_request(url: str, params: dict, access_token: str) -> dict:
+    """Make a request to the Meta Graph API."""
+    params["access_token"] = access_token
+    response = requests.get(url, params=params)
+
+    if response.status_code != 200:
+        raise Exception(f"Meta API request failed: {response.status_code} - {response.text}")
+
+    return response.json()
+
+
+def _make_paginated_api_request(
+    url: str, params: dict, access_token: str, time_range: dict | None = None
+) -> collections.abc.Generator[dict, None, None]:
+    """Make paginated requests to the Meta Graph API.
+    This function handles two types of pagination:
+    1. Standard pagination: Uses Meta's paging.next URLs to fetch all pages of results
+    2. Time-range pagination: Breaks large date ranges into monthly chunks to avoid slow API sorting,
+       then applies standard pagination within each monthly chunk
+    """
+    params["access_token"] = access_token
+
+    if time_range is None:
+        # Original pagination logic for non-time-range requests
+        next_url = url
+        while next_url:
+            if next_url == url:
+                response = requests.get(next_url, params=params)
+            else:
+                response = requests.get(next_url)
+
+            if response.status_code != 200:
+                raise Exception(f"Meta API request failed: {response.status_code} - {response.text}")
+
+            data = response.json()
+            yield data
+
+            paging = data.get("paging", {})
+            next_url = paging.get("next")
+    else:
+        start_date = dt.datetime.strptime(time_range["since"], "%Y-%m-%d")
+        end_date = dt.datetime.strptime(time_range["until"], "%Y-%m-%d")
+
+        current_start = start_date
```

Comment:
I would consider extracting out the time range logic out into its own function, so that it's a little less imbued in the paging and other parameter here. Perhaps easier to unit test outside the scope of an meta specific related concerns.